bindings/rust/s2n-tls/src/cert_chain.rs

```diff
@@ -6,34 +6,160 @@ use s2n_tls_sys::*;
 use std::{
     marker::PhantomData,
     ptr::{self, NonNull},
+    sync::Arc,
 };
 
+/// Internal wrapper type used for a convenient drop implementation.
+///
+/// [CertificateChain] is internally reference counted. The reference counted `T`
+/// must have a drop implementation.
+struct CertificateChainHandle {
+    cert: NonNull<s2n_cert_chain_and_key>,
+    is_owned: bool,
+}
+
+// # Safety
+//
+// s2n_cert_chain_and_key objects can be sent across threads.
+unsafe impl Send for CertificateChainHandle {}
+unsafe impl Sync for CertificateChainHandle {}
+
+impl CertificateChainHandle {
+    fn from_owned(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: true,
+        }
+    }
+
+    fn from_reference(cert: NonNull<s2n_cert_chain_and_key>) -> Self {
+        Self {
+            cert,
+            is_owned: false,
+        }
+    }
+}
+
+impl Drop for CertificateChainHandle {
+    fn drop(&mut self) {
+        // ignore failures since there's not much we can do about it
+        if self.is_owned {
+            unsafe {
+                let _ = s2n_cert_chain_and_key_free(self.cert.as_ptr()).into_result();
+            }
+        }
+    }
+}
+
+pub struct Builder {
+    cert: CertificateChain<'static>,
+}
+
+impl Builder {
+    pub fn new() -> Result<Self, Error> {
+        Ok(Self {
+            cert: CertificateChain::allocate_owned()?,
+        })
+    }
+
+    /// Corresponds to [s2n_cert_chain_and_key_load_pem_bytes]
+    ///
+    /// This can be used with [crate::config::Builder::load_chain] to share a
+    /// single cert across multiple configs.
+    pub fn load_pem(&mut self, chain: &[u8], key: &[u8]) -> Result<&mut Self, Error> {
+        unsafe {
+            // SAFETY: manual audit of load_pem_bytes shows that `chain_pem` and
+            // `private_key_pem` are not modified.
+            // https://github.com/aws/s2n-tls/issues/4140
+            s2n_cert_chain_and_key_load_pem_bytes(
+                self.cert.as_mut_ptr(),
+                chain.as_ptr() as *mut _,
+                chain.len() as u32,
+                key.as_ptr() as *mut _,
+                key.len() as u32,
+            )
+            .into_result()
+        }?;
+
+        Ok(self)
+    }
+
+    /// Corresponds to [s2n_cert_chain_and_key_load_public_pem_bytes].
+    ///
+    /// This method is only used when performing private-key offloading. For standard
+    /// use-cases see [CertificateChain::from_pem].
+    pub fn load_public_pem(&mut self, chain: &[u8]) -> Result<&mut Self, Error> {
+        unsafe {
+            // SAFETY: manual audit of load_public_pem_bytes shows that `chain_pem`
+            // is not modified
+            // https://github.com/aws/s2n-tls/issues/4140
+            s2n_cert_chain_and_key_load_public_pem_bytes(
+                self.cert.as_mut_ptr(),
+                chain.as_ptr() as *mut _,
+                chain.len() as u32,
+            )
+            .into_result()
```

Comment:
The same thing that happens if you call `config::Builder::load_pem` then `config::Builder::load_public_pem`. 😅 

I think the consistency is valuable, and this method preserves a bit more future flexibility (although I can't think of a scenario where you'd have a single certificate chain with both ...)

If we just want strong validation, what do you think about adding validation/errors on the builder? We could return an explicit error such as
> CertificateChain can only be associated with you certificate. You already called {method}, so it is invalid to call {method}`

Reply:
Consistency is a good argument :( You can't by definition have a chain that both has a private key and doesn't have a private key tho.

Not blocking, but it might actually be nice as a follow-up to add that validation error. It could also catch the case where we don't add any pem at all to the chain.